panda/src/event/pythonTask.cxx

```diff
@@ -175,8 +175,16 @@ get_args() {
     }
 
     this->ref();
-    PyObject *self = DTool_CreatePyInstance(this, Dtool_PythonTask, true, false);
-    PyTuple_SET_ITEM(with_task, num_args, self);
+
+    // Check whether we have a Python wrapper.  This is not the case if the
+    // object has been created by C++ and never been exposed to Python code.
+    if (__self__ == nullptr)
+    {
+      // A __self__ instance does not exist, let's create one now.
+      __self__ = DTool_CreatePyInstance(this, Dtool_PythonTask, true, false);
+    }
+
+    PyTuple_SET_ITEM(with_task, num_args, Py_XNewRef(__self__));
```

Comment:
This increments the refcount a second time if `__self__` was created. Or was that intended?

You can also use Py_NewRef, since it will never be nullptr

Reply:
Good question, I think we do need `__self__` to keep its initial reference and not have it stolen by the PyTuple_SET_ITEM call. The task could run multiple times via task.again and having the underlying `__self__` object being destructed before the task finishes could cause a premature clearing of everything cleared via tp_clear / `__clear__`.

No contesting on using Py_NewRef though, I'll fix that and the above style suggestion very soon.

Reply:
I've made some additional fixes -- I found that this was going to break the reference counting on the C++ side since my change was calling ref() even with we weren't making an additional DTool object instantiation. That has been fixed so we will only call `ref()` prior to creating  `__self__`. 

I also found that owner methods (e.g. _addTask/_clearTask) and the upon_death callbacks were not using the persistent python wrappers. That has been fixed and I've amended my test to check for correctness with both of those. I think this is good to go now unless there's any other issues